### PR TITLE
services/horizon: Add Horizon API support for CAP 33

### PR DESCRIPTION
**Pull request by tamirms**
<details>
  <summary>PR Checklist</summary>
  
### PR Structure

* [x] This PR has reasonably narrow scope (if not, break it down into smaller PRs).
* [x] This PR avoids mixing refactoring changes with feature changes (split into two PRs
  otherwise).
* [x] This PR's title starts with name of package that is most changed in the PR, ex.
  `services/friendbot`, or `all` or `doc` if the changes are broad or impact many
  packages.

### Thoroughness

* [x] This PR adds tests for the most critical parts of the new functionality or fixes.
* [ ] I've updated any docs ([developer docs](https://www.stellar.org/developers/reference/), `.md`
  files, etc... affected by this change). Take a look in the `docs` folder for a given service,
  like [this one](https://github.com/stellar/go/tree/master/services/horizon/internal/docs).

### Release planning

* [ ] I've updated the relevant CHANGELOG ([here](services/horizon/CHANGELOG.md) for Horizon) if
  needed with deprecations, added features, breaking changes, and DB schema changes.
* [ ] I've decided if this PR requires a new major/minor version according to
  [semver](https://semver.org/), or if it's mainly a patch change. The PR is targeted at the next
  release branch if it's not a patch change.
</details>

### What

Close https://github.com/stellar/go/issues/2862

Update Horizon responses to include sponsorship information. Also add endpoints for querying data by sponsorship.

- [x] Add attributes `num_sponsoring` and `num_sponsored` to accounts resource.

- [x] Add the attribute `sponsor to:

    - [x] `balances`
    - [x] `signers`
    - [x] `account`
    - [x] `account data `
        - [x] The sponsor will only be displayed when fetching the data object directly via /accounts/id/data/key
    - [x] `offers`
    - [x] `claimable_balances`

* Allow filtering by `sponsor` on the following resources:
    - [x] `/accounts?sponsor=accountID` returns all the accounts where the given `accountID` is sponsoring an account or any of its subentries.
    - [x] `/offers?sponsor=accountID` returns all offers where the given `accountID` is the sponsor.
    - [x] `/claimable_balances?sponsor=accountID` returns all claimable balances where the given `accountID` is the sponsor.

### Why

We need to extend the Horizon web API to expose CAP33 related data 

### Known limitations

[N/A]

### CODE REVIEW

File: protocols/horizon/main.go [hunk 1]
```diff
@@ -50,6 +50,9 @@ type Account struct {
 	Balances             []Balance         `json:"balances"`
 	Signers              []Signer          `json:"signers"`
 	Data                 map[string]string `json:"data"`
+	NumSponsoring        uint32            `json:"num_sponsoring"`
+	NumSponsored         uint32            `json:"num_sponsored"`
+	Sponsor              string            `json:"sponsor,omitempty"`
```
Comment (bartekn) [inline protocols/horizon/main.go:55]: I think `*string` and no `omitempty` make it more clear that it a ledger entry can have no sponsor, especially in `horizonclient`. Maybe let's discuss it quickly in Horizon meeting.

File: protocols/horizon/main.go [hunk 2]
```diff
@@ -137,23 +140,6 @@ func (a *Account) SignerSummary() map[string]int32 {
...
-	ID        string `json:"id"`
-	AccountID string `json:"account_id"`
-	PT        string `json:"paging_token"`
-	Signer    `json:"signer"`
-}
-
-// PagingToken implementation for hal.Pageable
-func (res AccountSigner) PagingToken() string {
-	return res.PT
-}
```
Comment (bartekn) [inline protocols/horizon/main.go:155]: Good catch 👍

File: services/horizon/internal/actions/account.go [hunk 3]
```diff
@@ -93,15 +94,18 @@ func (q AccountsQuery) Validate() error {
...
-		return invalidAccountsParams
+	numParams := 0
+	for _, present := range []bool{
+		len(q.Signer) > 0,
+		q.Asset() != nil,
+		len(q.Sponsor) > 0,
+	} {
+		if present {
+			numParams++
+		}
```
Comment (bartekn) [inline services/horizon/internal/actions/account.go:105]: We already have `countNonEmpty` for this (in `helpers.go`).

**Review by bartekn — APPROVED** [review]

File: services/horizon/internal/actions/account.go [hunk 4]
```diff
@@ -147,7 +151,12 @@ func (handler GetAccountsHandler) GetResourcePage(
 
 	var records []history.AccountEntry
 
-	if len(qp.Signer) > 0 {
+	if len(qp.Sponsor) > 0 {
+		records, err = historyQ.AccountsForSponsor(qp.Sponsor, pq)
```
Comment (abuiles) [inline services/horizon/internal/actions/account.go:155]: @tamirms this is returning only the account entries where `qp.Sponsor` is a sponsor, but it should return all the accounts where `qp.Sponsor` is sponsoring the account entry or any of its subentries (trust lines,  signers, data, or account entry)

**Review by abuiles — CHANGES_REQUESTED** [review]

**Review by abuiles — APPROVED** [review]

**Pull request merged**